File: wear/app/src/main/java/app/imuatrak/wear/services/StrokeDetector.kt
package app.imuatrak.wear.services

import kotlin.math.sqrt

// ── StrokeDetector.kt ─────────────────────────────────────────────────────────
// Direct Kotlin port of src/services/stroke-detector.ts.
// Same DSP coefficients — results are numerically consistent with the phone app.

data class Stroke(val tSec: Double, val rateSpm: Double, val confidence: Double)

class StrokeDetector(
    private val minSpm: Double = 30.0,
    private val maxSpm: Double = 120.0,
    private val peakThreshold: Double = 0.6,
) {
    private val refractorySec = 60.0 / maxSpm
    private val hpAlpha = 0.97
    private val lpAlpha = 0.25

    private var hpPrevIn = 0.0
    private var hpPrevOut = 0.0
    private var lpPrev = 0.0
    private var lastStrokeT = -1.0
    private var lastSampleT = -1.0
    private var lastValue = 0.0
    private var rising = false

    fun onSample(tSec: Double, ax: Double, ay: Double, az: Double): Stroke? {
        val magnitude = sqrt(ax*ax + ay*ay + az*az)
        val hp = hpAlpha * (hpPrevOut + magnitude - hpPrevIn)
        hpPrevIn = magnitude; hpPrevOut = hp
        val v = lpPrev + lpAlpha * (hp - lpPrev)
        lpPrev = v

        var stroke: Stroke? = null
        if (lastSampleT > 0) {
            val nowRising = v > lastValue
            if (rising && !nowRising && lastValue > peakThreshold) {
                if (lastStrokeT < 0 || tSec - lastStrokeT >= refractorySec) {
                    val rate = if (lastStrokeT < 0) 0.0 else 60.0 / (tSec - lastStrokeT)
                    if (rate == 0.0 || (rate in minSpm..maxSpm)) {
                        val prom = lastValue - peakThreshold
                        val conf = maxOf(0.0, minOf(1.0, prom / (prom + 0.5)))
                        stroke = Stroke(lastSampleT, rate, conf)
                        lastStrokeT = lastSampleT
                    }
                }
            }
            rising = nowRising
        }
        lastValue = v; lastSampleT = tSec
        return stroke
    }

    fun reset() {
        hpPrevIn = 0.0; hpPrevOut = 0.0; lpPrev = 0.0
        lastStrokeT = -1.0; lastSampleT = -1.0; lastValue = 0.0; rising = false
    }
}

File: apple-watch/Sources/Services/StrokeDetector.swift
import Foundation

// ── StrokeDetector.swift ──────────────────────────────────────────────────────
// Direct Swift port of src/services/stroke-detector.ts.
// Pipeline: |a| → high-pass → low-pass → peak pick with refractory window.
// Keep DSP coefficients identical so phone and watch sessions are comparable.

struct Stroke {
    let tSec: Double
    let rateSpm: Double
    let confidence: Double
}

final class StrokeDetector {
    private let minSpm: Double
    private let maxSpm: Double
    private let peakThreshold: Double
    private let refractorySec: Double

    // Same coefficients as TypeScript
    private let hpAlpha = 0.97
    private let lpAlpha = 0.25

    private var hpPrevIn = 0.0
    private var hpPrevOut = 0.0
    private var lpPrev = 0.0

    private var lastStrokeT = -1.0
    private var lastSampleT = -1.0
    private var lastValue = 0.0
    private var rising = false

    init(minSpm: Double = 30, maxSpm: Double = 120, peakThreshold: Double = 0.6) {
        self.minSpm = minSpm
        self.maxSpm = maxSpm
        self.peakThreshold = peakThreshold
        self.refractorySec = 60.0 / maxSpm
    }

    /// Feed one gravity-removed accelerometer sample. Returns a Stroke if a
    /// peak was detected, otherwise nil.
    func onSample(tSec: Double, ax: Double, ay: Double, az: Double) -> Stroke? {
        let magnitude = (ax*ax + ay*ay + az*az).squareRoot()

        let hp = hpAlpha * (hpPrevOut + magnitude - hpPrevIn)
        hpPrevIn = magnitude
        hpPrevOut = hp

        let v = lpPrev + lpAlpha * (hp - lpPrev)
        lpPrev = v

        var stroke: Stroke? = nil
        if lastSampleT > 0 {
            let nowRising = v > lastValue
            if rising && !nowRising && lastValue > peakThreshold {
                if lastStrokeT < 0 || tSec - lastStrokeT >= refractorySec {
                    let rate = lastStrokeT < 0 ? 0.0 : 60.0 / (tSec - lastStrokeT)
                    if rate == 0 || (rate >= minSpm && rate <= maxSpm) {
                        let prom = lastValue - peakThreshold
                        let conf = max(0, min(1, prom / (prom + 0.5)))
                        stroke = Stroke(tSec: lastSampleT, rateSpm: rate, confidence: conf)
                        lastStrokeT = lastSampleT
                    }
                }
            }
            rising = nowRising
        }
        lastValue = v
        lastSampleT = tSec
        return stroke
    }

    func reset() {
        hpPrevIn = 0; hpPrevOut = 0; lpPrev = 0
        lastStrokeT = -1; lastSampleT = -1; lastValue = 0; rising = false
    }
}
